File: app/src/main/java/com/cappielloantonio/tempo/ui/fragment/PlayerLyricsFragment.java
package com.cappielloantonio.tempo.ui.fragment;

import android.annotation.SuppressLint;
import android.content.ComponentName;
import android.os.Bundle;
import android.os.Handler;
import android.text.Layout;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.style.ForegroundColorSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.OptIn;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.media3.common.util.UnstableApi;
import androidx.media3.session.MediaBrowser;
import androidx.media3.session.SessionToken;

import com.cappielloantonio.tempo.R;
import com.cappielloantonio.tempo.databinding.InnerFragmentPlayerLyricsBinding;
import com.cappielloantonio.tempo.service.MediaService;
import com.cappielloantonio.tempo.subsonic.models.Line;
import com.cappielloantonio.tempo.subsonic.models.LyricsList;
import com.cappielloantonio.tempo.subsonic.models.StructuredLyrics;
import com.cappielloantonio.tempo.util.MusicUtil;
import com.cappielloantonio.tempo.util.Preferences;
import com.cappielloantonio.tempo.viewmodel.PlayerBottomSheetViewModel;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.android.material.button.MaterialButton;
import com.google.common.util.concurrent.MoreExecutors;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;


@OptIn(markerClass = UnstableApi.class)
public class PlayerLyricsFragment extends Fragment {
    private static final String TAG = "PlayerLyricsFragment";

    private InnerFragmentPlayerLyricsBinding bind;
    private PlayerBottomSheetViewModel playerBottomSheetViewModel;
    private ListenableFuture<MediaBrowser> mediaBrowserListenableFuture;
    private MediaBrowser mediaBrowser;
    private Handler syncLyricsHandler;
    private Runnable syncLyricsRunnable;
    private String currentLyrics;
    private LyricsList currentLyricsList;
    private String currentDescription;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        bind = InnerFragmentPlayerLyricsBinding.inflate(inflater, container, false);
        View view = bind.getRoot();

        playerBottomSheetViewModel = new ViewModelProvider(requireActivity()).get(PlayerBottomSheetViewModel.class);

        initOverlay();

        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        initPanelContent();
        observeDownloadState();
    }

    @Override
    public void onStart() {
        super.onStart();
        initializeBrowser();

    }

    @Override
    public void onResume() {
        super.onResume();
        bindMediaController();
        requireActivity().getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
    }

    @Override
    public void onPause() {
        super.onPause();
        releaseHandler();
        if (!Preferences.isDisplayAlwaysOn()) {
            requireActivity().getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
    }

    @Override
    public void onStop() {
        releaseBrowser();
        super.onStop();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        bind = null;
        currentLyrics = null;
        currentLyricsList = null;
        currentDescription = null;
    }

    private void initOverlay() {
        bind.syncLyricsTapButton.setOnClickListener(view -> {
            playerBottomSheetViewModel.changeSyncLyricsState();
        });

        bind.downloadLyricsButton.setOnClickListener(view -> {
            boolean saved = playerBottomSheetViewModel.downloadCurrentLyrics();
            if (getContext() != null) {
                Toast.makeText(
                        requireContext(),
                        saved ? R.string.player_lyrics_download_success : R.string.player_lyrics_download_failure,
                        Toast.LENGTH_SHORT
                ).show();
            }
        });
    }

    private void initializeBrowser() {
        mediaBrowserListenableFuture = new MediaBrowser.Builder(requireContext(), new SessionToken(requireContext(), new ComponentName(requireContext(), MediaService.class))).buildAsync();
    }

    private void releaseHandler() {
        if (syncLyricsHandler != null) {
            syncLyricsHandler.removeCallbacks(syncLyricsRunnable);
            syncLyricsHandler = null;
        }
    }

    private void releaseBrowser() {
        MediaBrowser.releaseFuture(mediaBrowserListenableFuture);
    }

    private void bindMediaController() {
        mediaBrowserListenableFuture.addListener(() -> {
            try {
                mediaBrowser = mediaBrowserListenableFuture.get();
                defineProgressHandler();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }, MoreExecutors.directExecutor());
    }

    private void initPanelContent() {
        playerBottomSheetViewModel.getLiveLyrics().observe(getViewLifecycleOwner(), lyrics -> {
            currentLyrics = lyrics;
            updatePanelContent();
        });

        playerBottomSheetViewModel.getLiveLyricsList().observe(getViewLifecycleOwner(), lyricsList -> {
            currentLyricsList = lyricsList;
            updatePanelContent();
        });

        playerBottomSheetViewModel.getLiveDescription().observe(getViewLifecycleOwner(), description -> {
            currentDescription = description;
            updatePanelContent();
        });
    }

    private void observeDownloadState() {
        playerBottomSheetViewModel.getLyricsCachedState().observe(getViewLifecycleOwner(), cached -> {
            if (bind != null) {
                MaterialButton downloadButton = (MaterialButton) bind.downloadLyricsButton;
                if (cached != null && cached) {
                    downloadButton.setIconResource(R.drawable.ic_done);
                    downloadButton.setContentDescription(getString(R.string.player_lyrics_downloaded_content_description));
                } else {
                    downloadButton.setIconResource(R.drawable.ic_download);
                    downloadButton.setContentDescription(getString(R.string.player_lyrics_download_content_description));
                }
            }
        });
    }

    private void updatePanelContent() {
        if (bind == null) {
            return;
        }

        bind.nowPlayingSongLyricsSrollView.smoothScrollTo(0, 0);

        if (hasStructuredLyrics(currentLyricsList)) {
            setSyncLirics(currentLyricsList);
            bind.nowPlayingSongLyricsTextView.setVisibility(View.VISIBLE);
            bind.emptyDescriptionImageView.setVisibility(View.GONE);
            bind.titleEmptyDescriptionLabel.setVisibility(View.GONE);
            bind.syncLyricsTapButton.setVisibility(View.VISIBLE);
            bind.downloadLyricsButton.setVisibility(View.VISIBLE);
            bind.downloadLyricsButton.setEnabled(true);
        } else if (hasText(currentLyrics)) {
            bind.nowPlayingSongLyricsTextView.setText(MusicUtil.getReadableLyrics(currentLyrics));
            bind.nowPlayingSongLyricsTextView.setVisibility(View.VISIBLE);
            bind.emptyDescriptionImageView.setVisibility(View.GONE);
            bind.titleEmptyDescriptionLabel.setVisibility(View.GONE);
            bind.syncLyricsTapButton.setVisibility(View.GONE);
            bind.downloadLyricsButton.setVisibility(View.VISIBLE);
            bind.downloadLyricsButton.setEnabled(true);
        } else if (hasText(currentDescription)) {
            bind.nowPlayingSongLyricsTextView.setText(MusicUtil.getReadableLyrics(currentDescription));
            bind.nowPlayingSongLyricsTextView.setVisibility(View.VISIBLE);
            bind.emptyDescriptionImageView.setVisibility(View.GONE);
            bind.titleEmptyDescriptionLabel.setVisibility(View.GONE);
            bind.syncLyricsTapButton.setVisibility(View.GONE);
            bind.downloadLyricsButton.setVisibility(View.GONE);
            bind.downloadLyricsButton.setEnabled(false);
        } else {
            bind.nowPlayingSongLyricsTextView.setVisibility(View.GONE);
            bind.emptyDescriptionImageView.setVisibility(View.VISIBLE);
            bind.titleEmptyDescriptionLabel.setVisibility(View.VISIBLE);
            bind.syncLyricsTapButton.setVisibility(View.GONE);
            bind.downloadLyricsButton.setVisibility(View.GONE);
            bind.downloadLyricsButton.setEnabled(false);
        }
    }

    private boolean hasText(String value) {
        return value != null && !value.trim().isEmpty();
    }

    private boolean listIsNotEmpty(List<?> list) {
        return list != null && !list.isEmpty();
    }

    private boolean hasStructuredLyrics(LyricsList lyricsList) {
        return Objects.nonNull(lyricsList)
                && listIsNotEmpty(lyricsList.getStructuredLyrics())
                && Objects.nonNull(lyricsList.getStructuredLyrics().get(0))
                && listIsNotEmpty(lyricsList.getStructuredLyrics().get(0).getLine());
    }

    @SuppressLint("DefaultLocale")
    private void setSyncLirics(LyricsList lyricsList) {
        if (lyricsList.getStructuredLyrics() != null && !lyricsList.getStructuredLyrics().isEmpty() && lyricsList.getStructuredLyrics().get(0).getLine() != null) {
            StringBuilder lyricsBuilder = new StringBuilder();
            List<Line> lines = lyricsList.getStructuredLyrics().get(0).getLine();

            if (lines != null) {
                for (Line line : lines) {
                    lyricsBuilder.append(line.getValue().trim()).append("\n");
                }
            }

            bind.nowPlayingSongLyricsTextView.setText(lyricsBuilder.toString());
        }
    }

    private void defineProgressHandler() {
        playerBottomSheetViewModel.getLiveLyricsList().observe(getViewLifecycleOwner(), lyricsList -> {
            if (!hasStructuredLyrics(lyricsList)) {
                releaseHandler();
                return;
            }

            if (!lyricsList.getStructuredLyrics().get(0).getSynced()) {
                releaseHandler();
                return;
            }

            syncLyricsHandler = new Handler();
            syncLyricsRunnable = () -> {
                if (syncLyricsHandler != null) {
                    if (bind != null) {
                        displaySyncedLyrics();
                    }

                    syncLyricsHandler.postDelayed(syncLyricsRunnable, 250);
                }
            };

            syncLyricsHandler.postDelayed(syncLyricsRunnable, 250);
        });
    }

    private void displaySyncedLyrics() {
        LyricsList lyricsList = playerBottomSheetViewModel.getLiveLyricsList().getValue();
        int timestamp = (int) (mediaBrowser.getCurrentPosition());

        if (hasStructuredLyrics(lyricsList)) {
            StringBuilder lyricsBuilder = new StringBuilder();
            List<Line> lines = lyricsList.getStructuredLyrics().get(0).getLine();

            if (lines == null || lines.isEmpty()) return;

            List<Line> curLines = getCurerntLyricsLine(lines, lyricsBuilder, timestamp);

            if (!curLines.isEmpty()) {
                Line toHighlight = curLines.get(0);
                String lyrics = lyricsBuilder.toString();
                Spannable spannableString = new SpannableString(lyrics);

                int startingPosition = getStartPosition(lines, toHighlight);
                int endingPosition = startingPosition + toHighlight.getValue().length();
                if (curLines.size() >1){
                    endingPosition = curLines.stream().mapToInt(line -> line.getValue().length() + 1).sum() + startingPosition -1;
                }

                spannableString.setSpan(new ForegroundColorSpan(requireContext().getResources().getColor(R.color.shadowsLyricsTextColor, null)), 0, lyrics.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                spannableString.setSpan(new ForegroundColorSpan(requireContext().getResources().getColor(R.color.lyricsTextColor, null)), startingPosition, endingPosition, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);

                bind.nowPlayingSongLyricsTextView.setText(spannableString);

                if (playerBottomSheetViewModel.getSyncLyricsState()) {
                    bind.nowPlayingSongLyricsSrollView.smoothScrollTo(0, getScroll(lines, toHighlight));
                }
            }
        }
    }

    @NonNull
    private List<Line> getCurerntLyricsLine(List<Line> lines, StringBuilder lyricsBuilder, int timestamp){
        lyricsBuilder.append("\n\n\n");
        int currentLineStart = 0;
        int firstIndex = 0, lastIndex = 0;
        for (int i = 0; i < lines.size(); i++) {
            Line line = lines.get(i);
            lyricsBuilder.append(line.getValue().trim()).append("\n");

            if (line.getStart() == null || lastIndex > 0){
                continue;
            }

            if (line.getStart() < timestamp) {
                if (currentLineStart < line.getStart()){
                    currentLineStart = line.getStart();
                    firstIndex = i;
                }
            }else{
                lastIndex = i;
            }
        }

        lyricsBuilder.append("\n\n\n\n\n"); // 额外追加一些换行，避免歌词末尾贴底。
        if (lastIndex == 0){
            if (firstIndex > 0){
                lastIndex = lines.size();
            }else{
                return new ArrayList<>(0);
            }
        }
        return lines.subList(firstIndex, lastIndex);
    }

    private int getStartPosition(List<Line> lines, Line toHighlight) {
        int start = 3;

        for (Line line : lines) {
            if (line != toHighlight) {
                start = start + line.getValue().length() + 1;
            } else {
                break;
            }
        }

        return start;
    }

    private int getLineCount(List<Line> lines, Line toHighlight) {
        int start = 0;

        for (Line line : lines) {
            if (line != toHighlight) {
                bind.tempLyricsLineTextView.setText(line.getValue());
                start = start + bind.tempLyricsLineTextView.getLineCount();
            } else {
                break;
            }
        }

        return start;
    }

    private int getScroll(List<Line> lines, Line toHighlight) {
        int startIndex = getStartPosition(lines, toHighlight);
        Layout layout = bind.nowPlayingSongLyricsTextView.getLayout();
        if (layout == null) return 0;

        int line = layout.getLineForOffset(startIndex);
        int lineTop = layout.getLineTop(line);
        int lineBottom = layout.getLineBottom(line);
        int lineCenter = (lineTop + lineBottom) / 2;

        int scrollViewHeight = bind.nowPlayingSongLyricsSrollView.getHeight();
        int scroll = lineCenter - scrollViewHeight / 2;

        return Math.max(scroll, 0);
    }
}

File: app/src/main/java/com/cappielloantonio/tempo/ui/fragment/PlayerLyricsFragment.kt
package com.cappielloantonio.tempo.ui.fragment

import android.content.ComponentName
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.media3.common.util.UnstableApi
import androidx.media3.session.MediaBrowser
import androidx.media3.session.SessionToken
import com.cappielloantonio.tempo.R
import com.cappielloantonio.tempo.databinding.InnerFragmentPlayerLyricsBinding
import com.cappielloantonio.tempo.service.MediaService
import com.cappielloantonio.tempo.subsonic.models.LyricsList
import com.cappielloantonio.tempo.ui.fragment.model.toBilingualLines
import com.cappielloantonio.tempo.util.MusicUtil
import com.cappielloantonio.tempo.util.Preferences
import com.cappielloantonio.tempo.viewmodel.PlayerBottomSheetViewModel
import com.google.common.util.concurrent.ListenableFuture
import com.google.common.util.concurrent.MoreExecutors

private const val TAG = "PlayerLyricsFragment"

@UnstableApi
class PlayerLyricsFragment : Fragment() {

    private var bind: InnerFragmentPlayerLyricsBinding? = null
    private lateinit var viewModel: PlayerBottomSheetViewModel

    private var browserFuture: ListenableFuture<MediaBrowser>? = null
    private var mediaBrowser: MediaBrowser? = null

    private val handler = Handler(Looper.getMainLooper())
    private val progressRunnable = object : Runnable {
        override fun run() {
            val browser = mediaBrowser
            val b = bind
            if (browser != null && b != null && browser.isConnected) {
                // 只有在播放时才频繁刷新，节省性能
                if (browser.isPlaying) {
                    b.lyricView.setProgress(browser.currentPosition)
                }
            }
            handler.postDelayed(this, 50)
        }
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        bind = InnerFragmentPlayerLyricsBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(requireActivity())[PlayerBottomSheetViewModel::class.java]
        return bind!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initButtons()
        observeLyrics()
        observeDownloadState()
    }

    override fun onStart() {
        super.onStart()
        browserFuture = MediaBrowser.Builder(
            requireContext(),
            SessionToken(requireContext(), ComponentName(requireContext(), MediaService::class.java))
        ).buildAsync()
    }

    override fun onResume() {
        super.onResume()
        bindMediaBrowser()
        requireActivity().window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    override fun onPause() {
        super.onPause()
        handler.removeCallbacks(progressRunnable)
        if (!Preferences.isDisplayAlwaysOn()) {
            requireActivity().window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }

    override fun onStop() {
        browserFuture?.cancel(true)
        mediaBrowser?.release()
        mediaBrowser = null
        super.onStop()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        bind = null
    }

    private fun initButtons() {
        bind?.syncLyricsTapButton?.setOnClickListener {
            viewModel.changeSyncLyricsState()
        }

        bind?.downloadLyricsButton?.setOnClickListener {
            val saved = viewModel.downloadCurrentLyrics()
            Toast.makeText(
                requireContext(),
                if (saved) R.string.player_lyrics_download_success else R.string.player_lyrics_download_failure,
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    // 在 bindMediaBrowser 成功后
    private fun bindMediaBrowser() {
        browserFuture?.addListener({
            try {
                mediaBrowser = browserFuture?.get()
                handler.removeCallbacks(progressRunnable) // 先移除防止重复
                handler.post(progressRunnable)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }, MoreExecutors.directExecutor())
    }

    private fun observeLyrics() {
        Log.d(TAG, "observeLyrics: ")
        viewModel.liveLyricsList.observe(viewLifecycleOwner) { lyricsList ->

            if (LyricsList.hasStructuredLyrics(lyricsList)) {
                val lyricLines = lyricsList.toBilingualLines()
                bind?.lyricView?.setLyrics(lyricLines)

                bind?.lyricView?.setOnLineClickListener { timeMs, _ ->
                    mediaBrowser?.seekTo(timeMs)
                }

                bind?.lyricView?.visibility = View.VISIBLE
                bind?.nowPlayingSongLyricsSrollView?.visibility = View.GONE
                bind?.nowPlayingSongLyricsTextView?.visibility = View.GONE
                bind?.emptyDescriptionImageView?.visibility = View.GONE
                bind?.titleEmptyDescriptionLabel?.visibility = View.GONE
//                目前没有拖拽歌词的功能，先把这个按钮直接隐藏掉
                bind?.syncLyricsTapButton?.visibility = View.GONE
                bind?.downloadLyricsButton?.visibility = View.VISIBLE
                bind?.downloadLyricsButton?.isEnabled = true

            } else {
                // fallback: plain text lyrics
                val text = viewModel.liveLyrics.value ?: viewModel.liveDescription.value
                if (!text.isNullOrBlank()) {
                    bind?.nowPlayingSongLyricsSrollView?.visibility = View.VISIBLE
                    bind?.nowPlayingSongLyricsTextView?.text = MusicUtil.getReadableLyrics(text)
                    bind?.nowPlayingSongLyricsTextView?.visibility = View.VISIBLE
                    bind?.lyricView?.visibility = View.GONE
                    bind?.emptyDescriptionImageView?.visibility = View.GONE
                    bind?.titleEmptyDescriptionLabel?.visibility = View.GONE
                    bind?.syncLyricsTapButton?.visibility = View.GONE
                    bind?.downloadLyricsButton?.visibility = View.GONE
                } else {
                    bind?.lyricView?.visibility = View.GONE
                    bind?.nowPlayingSongLyricsSrollView?.visibility = View.GONE
                    bind?.nowPlayingSongLyricsTextView?.visibility = View.GONE
                    bind?.emptyDescriptionImageView?.visibility = View.VISIBLE
                    bind?.titleEmptyDescriptionLabel?.visibility = View.VISIBLE
                    bind?.syncLyricsTapButton?.visibility = View.GONE
                    bind?.downloadLyricsButton?.visibility = View.GONE
                }
            }
        }
    }

    private fun observeDownloadState() {
        viewModel.lyricsCachedState.observe(viewLifecycleOwner) { cached ->
            val btn = bind?.downloadLyricsButton ?: return@observe
            if (cached == true) {
                btn.setIconResource(R.drawable.ic_done)
                btn.contentDescription = getString(R.string.player_lyrics_downloaded_content_description)
            } else {
                btn.setIconResource(R.drawable.ic_download)
                btn.contentDescription = getString(R.string.player_lyrics_download_content_description)
            }
        }
    }
}
